telemetry/src/metrics.rs

```diff
@@ -0,0 +1,174 @@
+//! [`Metrics`] and [`Status`]-related logic and functions.
+use std::{
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
+
+use prometheus::{
+    core::{AtomicU64, GenericGauge, GenericGaugeVec},
+    Encoder, Histogram, HistogramOpts, HistogramVec, IntCounter, IntCounterVec, Opts, Registry,
+};
+use serde::{Deserialize, Serialize};
+
+/// Thin wrapper around duration that `impl`s [`Default`]
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub struct Uptime(Duration);
+
+impl Default for Uptime {
+    fn default() -> Self {
+        Self(Duration::from_millis(0))
+    }
+}
+
+/// Response body for GET status request
+#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
+pub struct Status {
+    /// Number of connected peers, except for the reporting peer itself
+    pub peers: u64,
+    /// Number of committed blocks
+    pub blocks: u64,
+    /// Number of transactions committed in the last block
+    pub txs: u64,
+    /// Uptime since genesis block creation
+    pub uptime: Uptime,
+}
+
+impl From<&Arc<Metrics>> for Status {
```

Comment:
```suggestion
impl From<&Metrics> for Status {
```

I think deref coercion should take care of this

Reply:
Nope.

Reply:
ah, true. Then maybe:

```suggestion
impl<T: Deref<Target=Metrics>> From<T> for Status {
```